ce/var-con-it (#6)
Ada-Developers-Academy/ada-javascript-curriculum: 01-js-week-1/variables-conditionals-iteration.md

### We Can Do Declaration and Assignment Together

By the way, we can do declaration and initial assignment on one line:

```javascript
let clicheGoodbye = 'Don\'t forget to like and subscribe';
```

## Variables

> What's the syntax for defining a variable in this language? What's the syntax for accessing them?

> _-- Some cool programmer learning a new language_

In JavaScript, in order to define a variable, we do explicit variable declaration.

There are three keywords to use to declare a variable, all specifying different things:

1. `var`
1. `let`
1. `const`

Observe variable declaration and assignment below:

```javascript
var simonsFavColor = 'green';
let simonsFavNum = 72;
const simonsFavGame = 'night in the woods';
```

### We Prefer `let` for Reassignable Variables

`var` and `let` declare variables that can be re-assigned. `var` and `let` mean subtly different things, both having to do with [scope](https://stackoverflow.com/questions/762011/whats-the-difference-between-using-let-and-var-to-declare-a-variable) which we'll get to later.  **A good rule of thumb and our best practice moving forward is to always use `let` for variables we expect to be re-assigned.**

> `var` was the normal keyword to use in older versions of JavaScript. It may make sense to use `var` if you are working with an older version of the JavaScript interpreter. You may also see `var` when looking at old Stack Overflow posts.

Reviewer: This is worth a callout

```suggestion
<!-- available callout types: info, success, warning, danger, secondary  -->
### !callout-warning

## Avoid Using `var`

`var` was the normal keyword to use in older versions of JavaScript. It may make sense to use `var` if you are working with an older version of the JavaScript interpreter. You may also see `var` when looking at old Stack Overflow posts.

### !end-callout
```

### We Prefer `const` for All Other Variables

> Constants cannot change their value through reassignment. **We will always use `const` for variables we expect to not be re-assigned.** You should _always_ be using `const` by default, unless there's a good reason for that variable to be reassigned. Start every variable as `const`, and then refactor to `let` if you need to.

Reviewer: Maybe also turn this into a callout.

<!--BEGIN CHALLENGE-->

### !challenge

* type: multiple-choice
* id: 1e557275-f554-4477-bc22-30c5f200d708
* title: Predict the Output
<!--Other optional fields (checkpoints only) -->
<!--`points: 1`: the number of points for scoring as a checkpoint-->
<!--`topics: python, pandas`: the topics for analyzing points-->

##### !question

What is the output of the following code:
```javascript
favDogBreed;
let favDogBreed = "Newfoundland";
```

##### !end-question

##### !options

* undefined
* reference error
* "Newfoundland"

##### !end-options

##### !answer

reference error

##### !end-answer

<!--optional-->
##### !hint

##### !end-hint

<!--optional, checkpoints only-->
##### !rubric

##### !end-rubric

<!--optional-->
##### !explanation

##### !end-explanation

### !end-challenge

<!--END CHALLENGE-->
<!--BEGIN CHALLENGE-->

### !challenge

* type: multiple-choice
* id: 1aab9c44-e3dd-4f77-b6aa-9597ae9be411
* title: Predict the Output
<!--Other optional fields (checkpoints only) -->
<!--`points: 1`: the number of points for scoring as a checkpoint-->
<!--`topics: python, pandas`: the topics for analyzing points-->

##### !question

> What is the output of the following code:

Reviewer 2: ```suggestion
What is the output of the following `for` loop:
```

``` javascript
const temp = "3";
for(let i = 0; i <= 10; i++) {
  if(i == temp){
    return true;
  }
}

> return false;

Reviewer: Some might be confused that this isn't in a method or function.

```

##### !end-question

##### !options

* true
* false

##### !end-options

##### !answer

true

##### !end-answer

<!--optional-->
##### !hint

##### !end-hint

<!--optional, checkpoints only-->
##### !rubric

##### !end-rubric

<!--optional-->
##### !explanation

The == sign tests for loose equality. 3 and "3" are not strictly equal, but the == sign tries to convert values to a common data type *first* before comparing. Thus 3 and "3" returns **true**.

##### !end-explanation

### !end-challenge

<!--END CHALLENGE-->
<!--BEGIN CHALLENGE-->

### !challenge

* type: short-answer
* id: c629737a-1172-4c73-8044-efc01614970d
* title: Code Conversion
<!--Other optional fields (checkpoints only) -->
<!--`points: 1`: the number of points for scoring as a checkpoint-->
<!--`topics: python, pandas`: the topics for analyzing points-->

##### !question

Solve the following problem with a ternary operator: 
If the variable `input` is of data type string, return `true`; otherwise, return `false`.

##### !end-question

##### !answer

> ```javascript
> (typeof input === "string" ? true : false)
> ```

Reviewer: I don't think this works being put in the backticks like this.  

The answer expects a string or a regex expression.

```suggestion
/typeof input === [\"\']string[\"\'] \? true \: false/
```

Regardless, this will be a hard one for them to figure out by typing manually.  Maybe use a multiple-choice problem instead.
